pkg/didcomm/common/service/service.go

```diff
@@ -35,12 +35,39 @@ type Header struct {
 	Type   string           `json:"@type"`
 }
 
+func (h *Header) clone() *Header {
+	if h == nil {
+		return nil
+	}
+
+	return &Header{
+		ID: h.ID,
+		Thread: decorator.Thread{
+			ID: h.Thread.ID,
+		},
+		Type: h.Type,
+	}
+}
+
 // DIDCommMsg did comm msg
 type DIDCommMsg struct {
 	Header  *Header
 	Payload []byte
 }
 
+// Clone creates new DIDCommMsg with the same data
+// the cloned message is safe for delivering to the client
+// it prevents modifying by the client
+func (m *DIDCommMsg) Clone() *DIDCommMsg {
```

Comment:
`the cloned message is safe for delivering to the client` It means Client can not modify our data
without cloning, it can be a tricky hack how to change the protocol logic
e.g Client received an action event than he modified the message data and then he executes the callback (previously we stored msg in DB it was impossible to modify it)